qualtran/bloqs/chemistry/thc.ipynb

```diff
@@ -0,0 +1,302 @@
+    " - `num_spin_orb`: number of spin orbitals $N$\n",
+    " - `alt_mu`: Alternate values for mu indices.\n",
+    " - `alt_nu`: Alternate values for nu indices.\n",
+    " - `alt_theta`: Alternate values for theta indices.\n",
+    " - `theta`: Signs of lcu coefficients.\n",
+    " - `keep`: keep values.\n",
+    " - `keep_bitsize`: number of bits for keep register for coherent alias sampling. \n",
+    "\n",
+    "#### Registers\n",
+    " - `mu`: $\\mu$ register.\n",
+    " - `nu`: $\\nu$ register.\n",
+    " - `theta`: sign register.\n",
+    " - `succ`: success flag qubit from uniform state preparation\n",
+    " - `eq_nu_mp1`: flag for if $nu = M+1$\n",
+    " - `plus_a / plus_b`: plus state for controlled swaps on spins. \n",
+    "\n",
+    "#### References\n",
+    "[Even more efficient quantum computations of chemistry through tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf) Fig. 2 and Fig. 3.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c4886f09",
+   "metadata": {
+    "cq.autogen": "_make_prepare.py"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc import PrepareTHC\n",
+    "\n",
+    "num_spat = 4\n",
+    "num_mu = 8\n",
+    "t_l = np.random.normal(0, 1, size=num_spat)\n",
+    "zeta = np.random.normal(0, 1, size=(num_mu, num_mu))\n",
+    "zeta = 0.5 * (zeta + zeta.T)\n",
+    "eps = 1e-3\n",
+    "bloq = PrepareTHC.from_hamiltonian_coeffs(t_l, zeta, probability_epsilon=eps)\n",
+    "show_bloq(bloq)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e8c786a3",
+   "metadata": {},
+   "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ec8f8b91",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting import get_bloq_counts_graph\n",
+    "from qualtran.drawing import GraphvizCounts\n",
+    "from cirq_ft.algos.multi_control_multi_target_pauli import MultiControlPauli\n",
+    "\n",
+    "graph, sigma = get_bloq_counts_graph(bloq, generalizer=generalize)\n",
+    "GraphvizCounts(graph).get_svg()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "85818cf5",
+   "metadata": {},
+   "source": [
+    "### ```Paper Comparison```\n",
+    "\n",
+    "Let's compare our costs to those from the paper. Note we will only look at the cost of prepare. Inverting prepare has the same cost up to the cost of the inverse QROM being reduced to\n",
+    "\n",
+    "$$\n",
+    "\\lceil \\frac{d}{k_{s2}} \\rceil + k_{s2}\n",
+    "$$\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5be7dcf4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import bin_bloq_counts\n",
+    "\n",
+    "binned_counts = bin_bloq_counts(bloq)\n",
+    "data_size = bloq.num_mu * (bloq.num_mu + 1) // 2 + bloq.num_spin_orb // 2\n",
+    "\n",
+    "num_bits_mu = bloq.signature[0].bitsize\n",
```

Comment:
you use `bloq.num_mu` everywhere else so I know you know how :D

Reply:
Done.